tests/integration/utils.py

```diff
@@ -463,3 +465,166 @@ def _validate_expected_dep_file_contents(dep_contents_file: Path, output_dir: Pa
+    >>> is_testable_code(Path('README.md'))
+    False
+    >>> is_testable_code(Path('reqruiements.txt'))
+    False
+    """
+    return any(c.is_relative_to(p) for p in paths_to_code)
+
+
+def select_testable_changes(changes: tuple[Path, ...]) -> tuple[Path, ...]:
+    """Weed out changes that cannot be tested.
+
+    If a change is outside of paths to testable code it is dropped as if
+    it never happened. Any file or module outside of watched directories
+    will be rejected.
+    """
+    return tuple(c for c in changes if is_testable_code(c))
+
+
+def just_some_package_managers_were_affected_by(changes: tuple[Path, ...]) -> bool:
+    """Check that just package managers were affected.
+
+    If any code outside of package managers subtree was affected or if a module
+    shared by package managers was affected will return False.
+
+    >>> just_some_package_managers_were_affected_by((Path('tests/integration/test_pip.py'),))
+    True
+    >>> c = Path('cachi2/core/package_managers/pip.py'), Path('tests/integration/test_pip.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    True
+    >>> c = (Path('cachi2/core/package_managers/rpm/main.py'),)
+    >>> just_some_package_managers_were_affected_by(c)
+    True
+    >>> c = Path('cachi2/core/package_managers/gomod.py'), Path('tests/integration/test_pip.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    True
+    >>> c = Path('cachi2/core/package_managers/general.py'), Path('tests/integration/test_pip.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    False
+    >>> c = (Path('cachi2/core/package_managers/general.py'),
+    ...      Path('cachi2/core/package_managers/pip.py'))
+    >>> just_some_package_managers_were_affected_by(c)
+    False
+    >>> c = Path('tests/integration/utils.py'), Path('tests/integration/test_pip.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    False
+    >>> c = Path('cachi2/core/package_managers/pip.py'), Path('tests/integration/utils.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    False
+    >>> just_some_package_managers_were_affected_by((Path('cachi2/core/utils.py'),))
+    False
+    >>> c = (Path('cachi2/core/package_managers/general.py'),)
+    >>> just_some_package_managers_were_affected_by(c)
+    False
+    """
+    return all(affects_pm(c) for c in changes)
+
+
+def must_test_all() -> bool:
+    return os.getenv("CACHI2_RUN_ALL_INTEGRATION_TESTS", "false").lower() == "true"
+
+
+def determine_integration_tests_to_run_for(ref: str = "HEAD") -> Any:
+    """Check which tests to run basing on which files were changed in a commit."""
+    if must_test_all():
+        return set()
+    changes = select_testable_changes(retrieve_changed_files_from_git(ref))
+    if len(changes) == 0:
+        return supported_pms
+    elif just_some_package_managers_were_affected_by(changes):
+        return supported_pms - affected_package_managers(changes)
+    else:
+        return set()
```

Comment:
How about just `return set()` without the empty `else` clause at the end?

Reply:
I am not a big fan of dangling `elif`s, but yes, absolutely.